ddl/ddl_api.go

```diff
@@ -1369,6 +1372,109 @@ func checkPartitionByRangeColumn(ctx sessionctx.Context, tbInfo *model.TableInfo
+		if colInfo == nil {
+			return errors.Trace(ErrFieldNotFoundPart)
+		}
+		// The permitted data types are shown in the following list:
+		// All integer types
+		// DATE and DATETIME
+		// CHAR, VARCHAR, BINARY, and VARBINARY
+		// See https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.7/en/partitioning-columns.html
+		switch colInfo.FieldType.Tp {
+		case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong:
+		case mysql.TypeDate, mysql.TypeDatetime:
+		case mysql.TypeVarchar, mysql.TypeString:
+		default:
+			return ErrNotAllowedTypeInPartition.GenWithStackByArgs(col.O)
+		}
+	}
+	return nil
+}
+
+func checkRangeColumnsPartitionValue(ctx sessionctx.Context, tbInfo *model.TableInfo, pi *model.PartitionInfo) error {
+	// Range columns partition key supports multiple data types with integer、datetime、string.
+	defs := pi.Definitions
+	if len(defs) < 1 {
+		return errors.Trace(ErrPartitionsMustBeDefined)
+	}
+
+	curr := &defs[0]
+	if len(curr.LessThan) != len(pi.Columns) {
+		return errors.Trace(ErrPartitionColumnList)
+	}
+	for i := 1; i < len(defs); i++ {
+		prev, curr := curr, &defs[i]
+		succ, err := checkTwoRangeColumns(ctx, curr, prev, pi, tbInfo)
+		if err != nil {
+			return err
+		}
+		if !succ {
+			return errors.Trace(ErrRangeNotIncreasing)
+		}
+	}
+	return nil
+}
+
+func checkTwoRangeColumns(ctx sessionctx.Context, curr, prev *model.PartitionDefinition, pi *model.PartitionInfo, tbInfo *model.TableInfo) (bool, error) {
+	if len(curr.LessThan) != len(pi.Columns) {
+		return false, errors.Trace(ErrPartitionColumnList)
+	}
+	for i := 0; i < len(pi.Columns); i++ {
+		// Special handling for MAXVALUE.
+		if strings.EqualFold(curr.LessThan[i], partitionMaxValue) {
+			// If current is maxvalue, it certainly >= previous.
+			return true, nil
+		}
+		if strings.EqualFold(prev.LessThan[i], partitionMaxValue) {
+			// Current is not maxvalue, and previous is maxvalue.
+			return false, nil
+		}
+
+		// Current and previous is the same.
+		if strings.EqualFold(curr.LessThan[i], prev.LessThan[i]) {
+			continue
+		}
+
+		// The tuples of column values used to define the partitions are strictly increasing:
+		// PARTITION p0 VALUES LESS THAN (5,10,'ggg')
+		// PARTITION p1 VALUES LESS THAN (10,20,'mmm')
+		// PARTITION p2 VALUES LESS THAN (15,30,'sss')
+		succ, err := parseAndEvalBoolExpr(ctx, fmt.Sprintf("(%s) > (%s)", curr.LessThan[i], prev.LessThan[i]), tbInfo)
+		if err != nil {
+			return false, err
+		}
+
+		if succ {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func parseAndEvalBoolExpr(ctx sessionctx.Context, expr string, tbInfo *model.TableInfo) (bool, error) {
+	e, err := expression.ParseSimpleExprWithTableInfo(ctx, expr, tbInfo)
+	if err != nil {
+		return false, err
+	}
+	res, _, err1 := e.EvalInt(ctx, chunk.Row{})
+	if err1 != nil {
+		return false, err1
+	}
+	return res > 0, nil
+}
+
+func getColumnInfoByName(tbInfo *model.TableInfo, column string) *model.ColumnInfo {
```

Comment:
Could we use `FindCol` instead of this function?

Reply:
No. the parameter of `FindCol` is `[]*table.Column`, while the type of what `tbInfo.Cols()` returns is `[]*model.ColumnInfo`

Reply:
OK. Please put this function to ddl/column.go